src/ORM/DataObjectSchema.php

```diff
@@ -486,29 +486,76 @@ protected function cacheDatabaseFields($class)
     }
 
     /**
-     * Cache all indexes for the given class.
-     * Will do nothing if already cached
+     * Cache all indexes for the given class. Will do nothing if already cached.
      *
      * @param $class
      */
     protected function cacheDatabaseIndexes($class)
     {
-        if (array_key_exists($class, $this->databaseIndexes)) {
-            return;
+        if (!array_key_exists($class, $this->databaseIndexes)) {
+            $this->databaseIndexes[$class] = array_merge(
+                $this->defaultDatabaseIndexes($class),
+                $this->customDatabaseIndexes($class)
+            );
         }
-        $indexes = [];
+    }
+
+    /**
+     * Remove enum values from a type if they exist. If not, the type will be returned as is.
+     *
+     * @param  string $type E.g. DBClassName('SilverStripe\ORM\DataObject')
+     * @return string       E.g. DBClassName
+     */
+    protected function removeEnumValues($type)
+    {
+        $pieces = explode('(', $type);
+        return array_shift($pieces);
+    }
 
-        // look for indexable field types
-        foreach ($this->databaseFields($class, false) as $field => $type) {
-            if ($type === 'ForeignKey' || $type === 'DBClassName') {
+    /**
+     * Get "default" database indexable field types
+     *
+     * @param  string $class
+     * @return array
+     */
+    protected function defaultDatabaseIndexes($class)
+    {
+        $indexes = [];
+        $databaseFields = $this->databaseFields($class, false);
+        foreach ($databaseFields as $field => $type) {
```

Comment:
Would it be more efficient to simply loop over all has_one? At least you would have exactly one index per loop. Then you only need to compare the right-hand side of the has_one to check if it's polymorphic or not. I'd abandon the old implementation and start afresh.